nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol

```diff
@@ -16,43 +16,55 @@ contract Dispatcher is Upgradeable {
 
     /**
     * @param _target Target contract address
+    * @param _newSecretHash Secret hash (keccak256)
     **/
-    constructor(address _target) public {
+    constructor(address _target, bytes32 _newSecretHash) public {
         require(_target != 0x0);
         target = _target;
+        secretHash = _newSecretHash;
         require(target.delegatecall(bytes4(keccak256("finishUpgrade(address)")), target));
         emit Upgraded(0x0, _target, msg.sender);
     }
 
     /**
     * @notice Verify new contract storage and upgrade target
     * @param _target New target contract address
+    * @param _secret Secret for proof of contract owning
+    * @param _newSecretHash New secret hash (keccak256)
     **/
-    function upgrade(address _target) public onlyOwner {
+    function upgrade(address _target, bytes _secret, bytes32 _newSecretHash) public onlyOwner {
+        require(keccak256(_secret) == secretHash && _newSecretHash != secretHash);
```

Comment:
Wait.. Can _anyone_ who knows the secret upgrade?

That could (and will) be a problem if someone observes a transaction we broadcast and broadcasts a competing transaction. If you want to go this route, you need hash-and-reveal protocol.

(or did I understand not correctly, how it works?)

Reply:
When we thought this up, it was used as a second factor for upgrade verification.

In the event that someone can somehow upgrade the network with a stolen private key (or an exploit on ethereum), unless the attacker knows the secret, it won't upgrade.

Reply:
Ah, if it's only a _second_ factor, that is fine, yes

Reply:
Yeah, it's only second factor, the main factor is `onlyOwner`